### ISSUE TITLE
NFR: Get one element of the array with config->get()

### ISSUE THREAD

**Issue by xerron** [phalcon/cphalcon#12221]
The good from the get function is that you can set a default value.

now:

``` php
$adapter = $this->config->database->adapter ? $this->config->database->adapter : 'Mysql' ;
```

This is better:

``` php
$adapter = $this->config->get('database.adapter', 'Mysql');
```

**Comment by mzf** [comment]
Dot notation is very good idea

**Comment by sergeyklay** [comment]
I think will be better:
```php
$adapter = $this->config->path('application.database.adapter', 'Mysql');
$adapter = $this->config->path('application/database/adapter', 'Mysql', '/');
```

**Comment by mzf** [comment]
Ye. Fine option

**Comment by xerron** [comment]
it's a little confusing use 'path', for me is better 'get'

**Comment by Jurigag** [comment]
But what if for some weird reason someone already have named section with dot ?

**Comment by xerron** [comment]
@Jurigag  you mean this?

```php
<?php
return new \Phalcon\Config([
    'database' => [
        'adapter' => 'Postgresql',
        'host' => 'localhost',
        'username' => 'xerron',
        'password' => 'root',
        'dbname' => 'db_uwu'
    ],
   // ...code ...
    ' section.with.dot' =>  'mmm...'
]
```
good question. :+1:

**Comment by xerron** [comment]
@sergeyklay :+1:

**Comment by michanismus** [comment]
I need this so bad! 100x thumbs up

**Comment by sergeyklay** [comment]
Will be implemented in 3.1.0

**Comment by Jurigag** [comment]
So implement it in other places too, as far as i know people wanted this in forms too and somewhere else.

**Comment by xerron** [comment]
In PHP7 default values is most readable:

```php
$adapter =  $this->config->get('database.adapter') ?? 'Mysql';
```

**Comment by sergeyklay** [comment]
but it's longer and we support not only PHP7

**Comment by Jurigag** [comment]
Just add option to just call `$this->config->path('application.database.adapter')` without second argument, and you can use default value like php7 with  null coalescing operator.

**Comment by xerron** [comment]
Well, this is good:

```php
$adapter = $this->config->path('application/database/adapter', '/');
```

**Comment by daison12006013** [comment]
This is cool

**Comment by udarkness** [comment]
@sergeyklay  Actually what @xerron needs, is only:
$adapter = $this->config->database->adapter ?: 'Mysql' ;
Not PHP7 ?? required for this, just the classic Elvis operator ?:

**Comment by sergeyklay** [comment]
@udarkness Related to https://github.com/phalcon/cphalcon/issues/11157

**Comment by Yurist-85** [comment]
After read all above I think, the best way is something like this:

`$config = new Config([ ... ]);`
Use '.' [dot] notation by default but allowed to reset it if needed.
`$config->setPathSeparator('/');`

Use 3rd param to reset separator per request:
`$val = $config->get('some.other.path.to_the.val', null, '.');`

Return `null` if not exists (it's current behaviour I guess):
`$val = $config->get('some/path/to/nonexistantval');   // null`

Three ways to get default value:
```
$val = $config->get('some/path/to/nonexistantval', 'default_value');
$val = $config->get('some/path/to/nonexistantval') ?: 'default_value'; // 5.6
$val = $config->get('some/path/to/nonexistantval') ?? 'default_value'; // 7
```

And I'm absolutely **sure** that `->get()` is much more clear and obvious than `->path()` .

**Comment by fenixphp** [comment]
```php
get($key, $default = null)
```

---

```php
get('Dashboard/Main.plugins', [])
```

---

Up to the first dot, the path to the file. ``` Dashboard/Main ``` 
``` Dashboard/Main.json ``` or ``` Dashboard/Main.php ``` depending on the adapter.

After first dot using as array dot notation.

Something like this?

**Comment by Jurigag** [comment]
What? But config adapter itself accept file path, so no, it can't be like this.

**Comment by fenixphp** [comment]
А если он будет принимать путь к папке с конфигами, и получится ленивая загрузка конфигураций только там где это нужно и со значениями по умолчанию?

---

And if he will take the path to the configuration file, and you get lazy loading configurations only where it is needed and with default values?

---

Sorry translation may be inaccurate.

**Comment by sergeyklay** [comment]
Implemented in the `3.2.x` branch

**Comment by sergeyklay** [comment]
@fenixphp This should be a separated NFR
